src/Benchmark/IO/PointCloudIO.cpp

```diff
@@ -0,0 +1,174 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+// ----------------------------------------------------------------------------
+
+#include "benchmark/benchmark.h"
+
+#include "Open3D/IO/ClassIO/PointCloudIO.h"
+#include "Open3D/Utility/Console.h"
+
+namespace open3d {
+namespace benchmarks {
+
+using open3d::io::ReadPointCloud;
+using open3d::io::WritePointCloud;
+
+namespace {
+
+template <class T>
+double AverageDistance(const std::vector<T> &a, const std::vector<T> &b) {
+    if (a.size() != b.size()) {
+        utility::LogError("vectors different size {} {}", a.size(), b.size());
+    }
+    double total = 0;
+    for (size_t i = 0; i < a.size(); ++i) {
+        total += (a[i] - b[i]).norm();
+    }
+    return total / a.size();
+}
+
+enum class IsAscii : bool { BINARY = false, ASCII = true };
+enum class Compressed : bool { UNCOMPRESSED = false, COMPRESSED = true };
+enum class Compare : uint32_t {
+    // Points are always compared
+    NONE = 0,
+    NORMALS = 1 << 0,
+    COLORS = 1 << 1,
+    NORMALS_AND_COLORS = NORMALS | COLORS
+};
+struct ReadWritePCArgs {
+    std::string filename;
+    IsAscii write_ascii;
+    Compressed compressed;
+    Compare compare;
+};
+std::vector<ReadWritePCArgs> g_pc_args({
+        // PCD has ASCII, BINARY, and BINARY_COMPRESSED
+        {"testau.pcd", IsAscii::ASCII, Compressed::UNCOMPRESSED,
+         Compare::NORMALS_AND_COLORS},  // 0
+        {"testbu.pcd", IsAscii::BINARY, Compressed::UNCOMPRESSED,
+         Compare::NORMALS_AND_COLORS},  // 1
+        {"testbc.pcd", IsAscii::BINARY, Compressed::COMPRESSED,
+         Compare::NORMALS_AND_COLORS},  // 2
+        {"testb.ply", IsAscii::BINARY, Compressed::UNCOMPRESSED,
+         Compare::NORMALS_AND_COLORS},  // 3
+        {"testa.ply", IsAscii::ASCII, Compressed::UNCOMPRESSED,
+         Compare::NORMALS_AND_COLORS},  // 4
+        {"test.pts", IsAscii::BINARY, Compressed::UNCOMPRESSED,
+         Compare::COLORS},  // 5
+        {"test.xyz", IsAscii::BINARY, Compressed::UNCOMPRESSED,
+         Compare::NONE},  // 6
+        {"test.xyzn", IsAscii::BINARY, Compressed::UNCOMPRESSED,
+         Compare::NORMALS},  // 7
+        {"test.xyzrgb", IsAscii::BINARY, Compressed::UNCOMPRESSED,
+         Compare::COLORS},  // 8
+});
+
+class TestPCGrid0 {
+    geometry::PointCloud pc_;
+    int size_ = 0;
+
+public:
+    void Setup(int size) {
+        if (size_ == size) return;
+        utility::LogInfo("setup PCGrid size={}", size);
+        pc_.Clear();
+
+        size_ = size;
+        for (int i = 0; i < size; ++i) {
+            // provide somewhat random numbers everywhere, so compression
+            // doesn't get a free pass
+            pc_.points_.push_back({sin(i * .8969920581) * 1000.,
```

Comment:
shall we use `std::sin` and `std::fmod` to be more C++ish?

Reply:
updated